### ISSUE TITLE
The MPI_Abort() function was called after MPI_FINALIZE was invoked.

### ISSUE THREAD

**Issue by Tarkalak** [pghysels/STRUMPACK#25]
When I run a simple mpi program with strumpack I get the error bellow. That happens at the end of the program, when the MPI and BLACS stuff is closing.

I am on 

    $cat /etc/system-release-cpe 
    cpe:/o:redhat:enterprise_linux:7.7:ga:server

I have compiled STRUMPACK with OpenMPI version 4.0.2 and gcc 9.2.0.

[strumpack_simple_solver.tar.gz](https://github.com/pghysels/STRUMPACK/files/3964324/strumpack_simple_solver.tar.gz)

$ mpirun -np 4 ./main 250
0 0 values:
   0.001755 0.002898 0.002898 
   0.002021 0.003055 0.003793 
   0.001651 0.003699 0.004182 
0 0 sv alues:
   0.000000 
   0.000227 
   0.000874 
1 0 values:
   0.001333 0.002958 0.002958 
   0.001161 0.002424 0.004090 
   0.001038 0.002128 0.003365 
1 0 sv alues:
   0.001881 
   0.003179 
   0.004690 
0 1 values:
   0.004286 0.004700 0.004700 
   0.004352 0.004751 0.005017 
   0.004580 0.004918 0.005149 
1 1 values:
   0.005172 0.005261 0.005261 
   0.006489 0.006038 0.005841 
   0.005086 0.007656 0.006787 
*** The MPI_Abort() function was called after MPI_FINALIZE was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[nv001.avitohol.acad.bg:59551] Local abort after MPI_FINALIZE started completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
*** The MPI_Abort() function was called after MPI_FINALIZE was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[nv001.avitohol.acad.bg:59550] Local abort after MPI_FINALIZE started completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
*** The MPI_Abort() function was called after MPI_FINALIZE was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[nv001.avitohol.acad.bg:59552] Local abort after MPI_FINALIZE started completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
*** The MPI_Abort() function was called after MPI_FINALIZE was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[nv001.avitohol.acad.bg:59553] Local abort after MPI_FINALIZE started completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[39622,1],1]
  Exit code:    1
--------------------------------------------------------------------------

**Comment by pghysels** [comment]
Try putting everything enclosing everything from line 38 to line 112 in parentheses `{   }`.
Then the destructor for the `BLACSGrid` and the `DistributedMatrix` will be called before `Cblacs_exit` and `MPI_Finalize`.

**Comment by Tarkalak** [comment]
Hi Pieter,

Thank you, that works.

However I have another issue with the same program. I added solving a dense system with HSS and when I run it without `mpirun` or with `mpirun -np 1` it fails at
`        HSSMatrixMPI<double> H(A, hss_opts);`

with
`$ ./central 250
[nv001:81458] *** An error occurred in MPI_Group_incl
[nv001:81458] *** reported by process [1282277377,0]
[nv001:81458] *** on communicator MPI_COMM_WORLD
[nv001:81458] *** MPI_ERR_RANK: invalid rank
[nv001:81458] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[nv001:81458] ***    and potentially your MPI job)
`

Tests with more than mpi process run fine.

[strumpack_simple_solver.tar.gz](https://github.com/pghysels/STRUMPACK/files/3965297/strumpack_simple_solver.tar.gz)

**Comment by pghysels** [comment]
Yes, that is a known problem. It's not easy to fix. I will need to spend some time on that.
You can use HSSMatrix<double> instead. Sorry for the inconvenience.

**Comment by Tarkalak** [comment]
OK. Thanks.
